### PR TITLE
Edit tutorial to reset & record command buffers each frame

### LINKED ISSUE #202: Re-record command buffers every frame
While I understand that this tutorial is a beginners introduction into the API, many people work through it and then try to use it as the base for their endeavors in Vulkan. The issue is that outside of extremely basic tasks, it is difficult to use pre-recorded command buffers. Culling, adding & removing things to render every frame, as well as managing which swapchain image to render to are significantly simpler when you re-record every frame. I've seen countless beginners who've finished this tutorial series then are lost on how to implement features because of the lack of command buffer recording each frame.

What I propose is to rewrite the tutorial to re-record command buffers every frame instead of baking them ahead of time.

Obviously this is a very big undertaking. I am creating the issue to indicate that its something I consider a problem, and that if I created a PR which reworked the tutorials, it would be accepted (with as many revisions as necessary of course).

### PR DESCRIPTION
**Pull request by charles-lunarg**
This is a large change to the tutorial which makes the tutorial re-record command buffers. Originally described in #202 (and this PR fixes #202 )

Changes: 
* Modify the function that records into the command buffer so that it accepts a command buffer and image index.
* Split "Rendering and Presentation" into 2 chapters, one that introduces fences, semaphores, the main rendering functions (acquire/submit/present). Second chapter is dedicated to introducing multiple frames in flight.
* Rewrite the introduction of fences and semaphores.
* Made buffers, descriptors, and command buffers have MAX_IMAGE_IN_FLIGHT duplication, not swapchain image count. 
* Remove command buffer, descriptor set, and `VkBuffer` recreation from swapchain resizing
* Remove `imagesInFlight`, since it is no longer necessary.

### CODE REVIEW

**Comment by Overv** [comment]
Very nice! I'll be able to start reviewing this PR sometime this week.

**Comment by charles-lunarg** [comment]
I has come to my attention that there may be a sync bug W.R.T the depth buffer. Gist is that the same depth buffer could be read/written to by multiple frames in flight. A much more detailed description & solution is presented in the vk-guide.dev repo.
https://github.com/vblanco20-1/vulkan-guide/pull/53

**Comment by ralphtheninja** [comment]
@charles-lunarg Thanks for taking your time with this. I'm a beginner myself and I was pondering these exact concepts the other day.

File: code/20_staging_buffer.cpp [hunk 1]
```diff
@@ -253,7 +250,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/20_staging_buffer.cpp:254]:
```suggestion
```

File: code/17_swap_chain_recreation.cpp [hunk 2]
```diff
@@ -206,7 +203,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/17_swap_chain_recreation.cpp:207]:
```suggestion
```

File: code/21_index_buffer.cpp [hunk 3]
```diff
@@ -264,7 +261,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/21_index_buffer.cpp:265]:
```suggestion
```

File: code/22_descriptor_layout.cpp [hunk 4]
```diff
@@ -286,8 +283,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createUniformBuffers();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/22_descriptor_layout.cpp:287]:
```suggestion
```

File: code/22_descriptor_layout.cpp [hunk 5]
```diff
@@ -940,12 +936,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/22_descriptor_layout.cpp:939]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/23_descriptor_sets.cpp [hunk 6]
```diff
@@ -293,10 +290,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/23_descriptor_sets.cpp:294]:
```suggestion
```

File: code/23_descriptor_sets.cpp [hunk 7]
```diff
@@ -999,12 +994,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/23_descriptor_sets.cpp:997]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/24_texture_image.cpp [hunk 8]
```diff
@@ -303,10 +300,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/24_texture_image.cpp:304]:
```suggestion
```

File: code/24_texture_image.cpp [hunk 9]
```diff
@@ -1154,12 +1149,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/24_texture_image.cpp:1152]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/25_sampler.cpp [hunk 10]
```diff
@@ -310,10 +307,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/25_sampler.cpp:311]:
```suggestion
```

File: code/25_sampler.cpp [hunk 11]
```diff
@@ -1193,12 +1187,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/25_sampler.cpp:1190]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/26_texture_mapping.cpp [hunk 12]
```diff
@@ -316,10 +313,6 @@ class HelloTriangleApplication {
         createRenderPass();
         createGraphicsPipeline();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/26_texture_mapping.cpp:317]:
```suggestion
```

File: code/26_texture_mapping.cpp [hunk 13]
```diff
@@ -1223,12 +1217,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/26_texture_mapping.cpp:1220]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/27_depth_buffering.cpp [hunk 14]
```diff
@@ -333,10 +330,6 @@ class HelloTriangleApplication {
         createGraphicsPipeline();
         createDepthResources();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/27_depth_buffering.cpp:334]:
```suggestion
```

File: code/27_depth_buffering.cpp [hunk 15]
```diff
@@ -1303,12 +1297,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/27_depth_buffering.cpp:1300]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/28_model_loading.cpp [hunk 16]
```diff
@@ -340,10 +337,6 @@ class HelloTriangleApplication {
         createGraphicsPipeline();
         createDepthResources();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/28_model_loading.cpp:341]:
```suggestion
```

File: code/28_model_loading.cpp [hunk 17]
```diff
@@ -1349,12 +1343,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/28_model_loading.cpp:1346]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/29_mipmapping.cpp [hunk 18]
```diff
@@ -341,10 +338,6 @@ class HelloTriangleApplication {
         createGraphicsPipeline();
         createDepthResources();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/29_mipmapping.cpp:342]:
```suggestion
```

File: code/29_mipmapping.cpp [hunk 19]
```diff
@@ -1443,12 +1437,11 @@ class HelloTriangleApplication {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/29_mipmapping.cpp:1440]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

File: code/30_multisampling.cpp [hunk 20]
```diff
@@ -352,10 +349,6 @@ class HelloTriangleApplication {
         createColorResources();
         createDepthResources();
         createFramebuffers();
-        createUniformBuffers();
-        createDescriptorPool();
-        createDescriptorSets();
-        createCommandBuffers();
 
         imagesInFlight.resize(swapChainImages.size(), VK_NULL_HANDLE);
```
Comment (Overv) [inline code/30_multisampling.cpp:353]:
```suggestion
```

File: code/30_multisampling.cpp [hunk 21]
```diff
@@ -1493,12 +1487,11 @@ VkSampleCountFlagBits getMaxUsableSampleCount() {
             throw std::runtime_error("failed to acquire swap chain image!");
         }
 
-        updateUniformBuffer(imageIndex);
+        updateUniformBuffer(currentFrame);
```
Comment (Overv) [inline code/30_multisampling.cpp:1490]:
```suggestion
        updateUniformBuffer((uint32_t) currentFrame);
```

**Review by Overv — COMMENTED** [review]
Awesome work! I have only some minor comments 👍

**Comment by ralphtheninja** [comment]
Nice review!

**Comment by Overv** [comment]
@Krenodeno Would you be interested in incorporating these changes in the French translation as well? Otherwise we'll have to add a disclaimer that some parts of the translation are out of date.

**Comment by Krenodeno** [comment]
Sure, I'll do it. Not sure how much time it will take though, I'll do my best :)

File: en/03_Drawing_a_triangle/03_Drawing/02_Rendering_and_presentation.md [hunk 22]
```diff
@@ -19,46 +19,162 @@ void drawFrame() {
...
+A fence has a similar purpose, in that it is used to synchronize execution, but
+it is for ordering the execution on the CPU, otherwise known as the host.
+Simply put, if the host needs to know when the GPU has finished something, we
+use a fence.
+
+Similar to semaphores, fences can be unsignaled or signaled. Whenever we submit
+work to execute, we can attach a fence to that work. When the work is finished,
+the fence will be signaled. This can be waited upon by the host. Note that we
+manually reset fences, unlike semaphores which are automatically reset after
+being waited upon.
```
[the remark this answers is not in the document]
Comment (charles-lunarg) [reply en/03_Drawing_a_triangle/03_Drawing/02_Rendering_and_presentation.md]: Agreed, expanded the description of fence & semaphore resetting.

File: en/03_Drawing_a_triangle/03_Drawing/03_Frames_in_flight.md [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+of the next. How do we do this? Any resource that is accessed and modified
+during rendering must be duplicated. Thus, we need multiple command buffers,
+semaphores, and fences. In later chapters we will also add multiple instances
+of other resources, so we will see concept reappear.
+
+Start by adding a constant at the top of the program that defines how many
+frames should be processed concurrently:
+
+```c++
+const int MAX_FRAMES_IN_FLIGHT = 2;
```
[the remark this answers is not in the document]
Comment (charles-lunarg) [reply en/03_Drawing_a_triangle/03_Drawing/03_Frames_in_flight.md:20]: Definitely. Added a paragraph explaining how this allows the CPU and GPU to do work at the same time, and that making the number bigger can add frames of latency. Generally, this isn't desired but vulkan is explicit so it gives us the choice.

File: code/22_descriptor_layout.cpp (hunk 5 above)
Comment (charles-lunarg) [reply code/22_descriptor_layout.cpp:936]: I changed the type of currentFrame to be `uint32_t` so no casting is required. Seemed simpler than adding casts everywhere to make the compiler happy.

**Comment by charles-lunarg** [comment]
I addressed all the changes requested by Overv, cleaning up some paragraphs, and fixing resizing to not deadlock.

These include:
* Adding more information about fence and semaphore resetting
* Adding a paragraph explaining the choice of 2 for MAX_FRAMES_IN_FLIGHT
* Using uint32_t for `currentFrame` instead of size_t
* Moving resetting the fence to *after* acquiring a swapchain image
* Uncommenting some accidentally commented code.

File: en/03_Drawing_a_triangle/03_Drawing/03_Frames_in_flight.md [hunk 24]
```diff
@@ -20,8 +20,16 @@ frames should be processed concurrently:
 const int MAX_FRAMES_IN_FLIGHT = 2;
 ```
 
+We choose the number 2 because we don't want the CPU to get *too* far ahead of
+the GPU. With 2 frames in flight, the CPU and the GPU can be working on their
+own tasks at the same time. If the CPU finishes early, it will wait till the
+GPU finishes rendering before submitting more work. With 3 or more frames in
+flight, the CPU could get ahead of the GPU, adding frames of latency.
+Generally, extra latency isn't desired. But giving the application control over
+the number of frames in flight is another example of Vulkan being explicit.
```
Comment (ralphtheninja) [inline en/03_Drawing_a_triangle/03_Drawing/03_Frames_in_flight.md:29]: :+1:

**Review by Overv — APPROVED** [review]

**Comment by Krenodeno** [comment]
@Overv How can I edit the translation ? Forking the fork will not display change here (or does  it ?), or Charles will have to merge to be displayed here.

**Comment by Overv** [comment]
> @Overv How can I edit the translation ? Forking the fork will not display change here (or does it ?), or Charles will have to merge to be displayed here.

There are still going to be a few changes to this PR (possibly somewhat significant), so I'd wait a bit with the translation work. I think that once this PR is finished it would be good to get the changes in the English version out as quickly as possible and then you can work on the translation in your own pull request just like with earlier changes.

**Comment by Krenodeno** [comment]
> There are still going to be a few changes to this PR (possibly somewhat significant), so I'd wait a bit with the translation work. I think that once this PR is finished it would be good to get the changes in the English version out as quickly as possible and then you can work on the translation in your own pull request just like with earlier changes.

Okay, that's good for me !

**Comment by charles-lunarg** [comment]
FYI There is nothing more to do in this PR as far as I can tell.
I don't have any outstanding questions or concerns. Totally capable of doing further revisions if someone else wants to review.

**Comment by SaschaWillems** [comment]
Totally happy with these changes. They make the tutorial as a whole much better. Would love to se this merged and the website updated :)

**Comment by Overv** [comment]
Excellent work!

**Comment by ralphtheninja** [comment]
@Overv Is this automatically deployed to the live tutorial?

**Comment by Overv** [comment]
@ralphtheninja Yeah, the website is automatically redeployed once a day.

**Comment by Overv** [comment]
@Krenodeno Would you have time to work on the translation soon? :)

**Comment by Krenodeno** [comment]
I'll try to look at it this week. 
I think there are other changes made by other PRs that I'll have to update as well.

**Pull request merged**